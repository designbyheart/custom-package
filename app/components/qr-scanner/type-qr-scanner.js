// @flow

export const SCAN_STATUS = {
  SCANNING: 'scanning...',
  SUCCESS: 'Success!',
  FAIL: 'Failed to scan QR code',
}

export type QrCode = {
  id: string,
  s: {
    n: string,
    dp: {
      d: string,
      k: string,
      s: string,
    },
    d: string,
    l: string,
    v: string,
  },
  sa: {
    d: string,
    v: string,
    e: string,
  },
  t: string,
}

type ValuesType = <V>(v: V) => V

type QR_SCAN_STATUS =
  | typeof SCAN_STATUS.SCANNING
  | typeof SCAN_STATUS.FAIL
  | typeof SCAN_STATUS.SUCCESS

export type QrScannerState = {
  scanning: boolean,
  scanStatus: QR_SCAN_STATUS,
}

export type QrScannerProps = {
  onRead: QrCode => void,
  onClose: any => void,
}
