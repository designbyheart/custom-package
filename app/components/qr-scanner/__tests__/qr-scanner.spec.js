// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { SCAN_STATUS } from '../type-qr-scanner'
import QRScanner from '../qr-scanner'

describe('<QRScanner />', () => {
  const props = () => ({
    onClose: jest.fn(),
    onRead: jest.fn(),
  })

  it('should match snapshot', () => {
    const { onClose, onRead } = props()
    const wrapper = renderer
      .create(<QRScanner onClose={onClose} onRead={onRead} />)
      .toJSON()
    expect(wrapper).toMatchSnapshot()
  })

  it('should call onRead once QR code read is successful', () => {
    jest.useFakeTimers()
    const { onClose, onRead } = props()
    const instance = renderer
      .create(<QRScanner onClose={onClose} onRead={onRead} />)
      .getInstance()

    const qrData = {
      lu: 'http://logourl.com',
      rid: 'requestid',
      sakdp: 'senderAgentKeyDlgProof',
      sn: 'senderName',
      tn: 'targetName',
      sD: 'senderDID',
      sVk: 'senderVerificationKey',
      e: 'https://remoteendpoint.com/agent',
    }
    const qrReadEvent = {
      data: JSON.stringify(qrData),
    }

    instance.onRead(qrReadEvent)
    expect(onRead).toHaveBeenCalledWith(expect.objectContaining(qrData))
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SUCCESS)

    jest.runAllTimers()
    // qr scanner should be reactivated after all timers are done
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  it('should set state to fail if QR code is not correct', () => {
    jest.useFakeTimers()

    const { onClose, onRead } = props()
    const instance = renderer
      .create(<QRScanner onClose={onClose} onRead={onRead} />)
      .getInstance()

    instance.onRead({ data: '' })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.FAIL)

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })
})
