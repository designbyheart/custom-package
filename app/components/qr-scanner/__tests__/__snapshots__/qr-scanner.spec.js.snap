// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QRScanner /> should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "SF Pro Display",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Scan QR Code
      </Text>
      <View
        spaceBetween={true}
        style={
          Object {
            "height": 250,
            "justifyContent": "space-between",
            "marginVertical": 20,
            "width": 250,
          }
        }
        transparentBg={true}
      >
        <View
          row={true}
          spaceBetween={true}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "borderColor": "#ffffff",
                "borderLeftWidth": 5,
                "borderTopWidth": 5,
                "height": 50,
                "width": 50,
              }
            }
            transparentBg={true}
          />
          <View
            style={
              Object {
                "borderColor": "#ffffff",
                "borderRightWidth": 5,
                "borderTopWidth": 5,
                "height": 50,
                "width": 50,
              }
            }
            transparentBg={true}
          />
        </View>
        <View
          row={true}
          spaceBetween={true}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "borderBottomWidth": 5,
                "borderColor": "#ffffff",
                "borderLeftWidth": 5,
                "height": 50,
                "width": 50,
              }
            }
            transparentBg={true}
          />
          <View
            style={
              Object {
                "borderBottomWidth": 5,
                "borderColor": "#ffffff",
                "borderRightWidth": 5,
                "height": 50,
                "width": 50,
              }
            }
            transparentBg={true}
          />
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "SF Pro Display",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "color": "#ffffff",
            },
          ]
        }
      >
        scanning...
      </Text>
      <View
        center={true}
        row={true}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginVertical": 20,
          }
        }
        testID="close-qr-scanner-container"
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              undefined,
            ]
          }
          testID="close-qr-scanner-button"
          tvParallaxProperties={undefined}
        >
          <Image
            onPress={[Function]}
            source={1}
            testID="close-qr-scanner-icon"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.2)",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": -1,
          }
        }
      />
    </View>
  </View>
</View>
`;
